File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/data/model/book/Book.java
package ru.kmorozov.library.data.model.book;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.TextIndexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Created by sbt-morozov-kv on 28.11.2016.
 */

@Document
public class Book {

    @Id
    String bookId;
    @TextIndexed
    String title;
    @TextIndexed
    String author;

    BookInfo bookInfo;

    LinkInfo linkInfo;

    @DBRef
    Storage storage;

    @DBRef(lazy = true)
    Set<Category> categories;

    Map<IdType, String> bookIds;

    public Book() {
    }

    public Map<IdType, String> getBookIds() {
        return bookIds;
    }

    public void setBookIds(final Map<IdType, String> bookIds) {
        this.bookIds = bookIds;
    }

    public Book(String title, String author) {
        this.title = title;
        this.author = author;
    }

    public String getBookId() {
        return this.bookId;
    }

    public void setBookId(String bookId) {
        this.bookId = bookId;
    }

    public String getTitle() {
        return this.title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAuthor() {
        return this.author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public BookInfo getBookInfo() {
        return this.bookInfo;
    }

    public void setBookInfo(BookInfo bookInfo) {
        this.bookInfo = bookInfo;
    }

    public Storage getStorage() {
        return this.storage;
    }

    public void setStorage(Storage storage) {
        this.storage = storage;
    }

    public String getBookKey() {
        return this.bookInfo.getFileName();
    }

    public LinkInfo getLinkInfo() {
        return this.linkInfo;
    }

    public void setLinkInfo(LinkInfo linkInfo) {
        this.linkInfo = linkInfo;
    }

    public boolean isLink() {
        return bookInfo.isLink();
    }

    public boolean isBrokenLink() {
        return bookInfo.isLink() && ((null == linkInfo || linkInfo.isBroken()));
    }

    public Set<Category> getCategories() {
        return categories;
    }

    public void setCategories(final Set<Category> categories) {
        this.categories = categories;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (null == o || this.getClass() != o.getClass()) return false;
        if (null == this.title) return false;

        Book book = (Book) o;

        if (!this.title.equals(book.title)) return false;
        return this.author.equals(book.author);
    }

    @Override
    public int hashCode() {
        if (null == this.title)
            return super.hashCode();

        int result = this.title.hashCode();
        result = 31 * result + this.author.hashCode();
        return result;
    }

    public void addBookId(final IdType idType, final String bookId) {
        this.bookIds = this.bookIds == null ? new HashMap<>(1) : this.bookIds;
        this.bookIds.put(idType, bookId);
    }
}

File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/data/model/book/Book.kt
package ru.kmorozov.library.data.model.book

import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.TextIndexed
import org.springframework.data.mongodb.core.mapping.DBRef
import org.springframework.data.mongodb.core.mapping.Document

import java.util.HashMap

/**
 * Created by sbt-morozov-kv on 28.11.2016.
 */

@Document
class Book {

    @Id
    lateinit var bookId: String
    @TextIndexed
    var title: String? = null
    @TextIndexed
    lateinit var author: String

    lateinit var bookInfo: BookInfo

    var linkInfo: LinkInfo? = null

    @DBRef
    var storage: Storage? = null

    @DBRef(lazy = true)
    var categories: Set<Category>? = null

    internal var bookIds: MutableMap<IdType, String>? = null

    val bookKey: String?
        get() = bookInfo.fileName

    val isLink: Boolean
        get() = this.bookInfo.isLink

    val isBrokenLink: Boolean
        get() = this.bookInfo.isLink && (null == this.linkInfo || this.linkInfo!!.isBroken)

    constructor()

    fun getBookIds(): Map<IdType, String>? {
        return this.bookIds
    }

    fun setBookIds(bookIds: MutableMap<IdType, String>) {
        this.bookIds = bookIds
    }

    constructor(title: String, author: String) {
        this.title = title
        this.author = author
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (null == other || javaClass != other.javaClass) return false
        if (null == title) return false

        val book = other as Book?

        return if (title != book!!.title) false else author == book.author
    }

    override fun hashCode(): Int {
        if (null == title)
            return super.hashCode()

        var result = title!!.hashCode()
        result = 31 * result + author.hashCode()
        return result
    }

    fun addBookId(idType: IdType, bookId: String) {
        bookIds = if (bookIds == null) HashMap(1) else bookIds
        bookIds!![idType] = bookId
    }
}
